crates/sui-adapter/src/programmable_transactions/linkage_view.rs

```diff
@@ -0,0 +1,258 @@
+                        defining_id,
+                        entry.get(),
+                    )));
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl From<&MovePackage> for LinkageInfo {
+    fn from(package: &MovePackage) -> Self {
+        Self {
+            link_context: package.id().into(),
+            original_package_id: package.original_package_id().into(),
+            linkage_table: package.linkage_table().clone(),
+        }
+    }
+}
+
+impl<'state, S: StorageView> LinkageResolver for LinkageView<'state, S> {
+    type Error = SuiError;
+
+    fn link_context(&self) -> AccountAddress {
+        if let Some(LinkageInfo { link_context, .. }) = &self.linkage_info {
+            *link_context
+        } else {
+            AccountAddress::ZERO
+        }
+    }
+
+    fn relocate(&self, module_id: &ModuleId) -> Result<ModuleId, Self::Error> {
+        let Some(linkage) = &self.linkage_info else {
+            return Ok(module_id.clone());
+        };
+
+        // The request is to relocate a module in the package that the link context is from.  This
+        // entry will not be stored in the linkage table, so must be handled specially.
+        if module_id.address() == &linkage.original_package_id {
+            return Ok(ModuleId::new(
+                linkage.link_context,
+                module_id.name().to_owned(),
+            ));
+        }
+
+        let runtime_id = ObjectID::from_address(*module_id.address());
+        let Some(upgrade) = linkage.linkage_table.get(&runtime_id) else {
+            return Err(ExecutionError::invariant_violation(format!(
+                "Missing linkage for {runtime_id} in context {}",
+                linkage.link_context,
+            )).into());
+        };
+
+        Ok(ModuleId::new(
+            upgrade.upgraded_id.into(),
+            module_id.name().to_owned(),
+        ))
+    }
+
+    fn defining_module(
+        &self,
+        runtime_id: &ModuleId,
+        struct_: &IdentStr,
+    ) -> Result<ModuleId, Self::Error> {
+        if self.linkage_info.is_none() {
+            return Ok(runtime_id.clone());
+        }
+
+        if let Some(cached) = self.get_cached_type_origin(runtime_id, struct_) {
+            return Ok(ModuleId::new(cached, runtime_id.name().to_owned()));
+        }
+
+        let storage_id = ObjectID::from(*self.relocate(runtime_id)?.address());
+        let Some(package) = self.state_view.get_package(&storage_id)? else {
+            return Err(ExecutionError::invariant_violation(format!(
+                "Missing dependent package in store: {storage_id}",
+            )).into())
+        };
+
+        for TypeOrigin {
+            module_name,
+            struct_name,
+            package,
+        } in package.type_origin_table()
+        {
+            if module_name == runtime_id.name().as_str() && struct_name == struct_.as_str() {
+                self.add_type_origin(runtime_id.clone(), struct_.to_owned(), *package)?;
```

Comment:
Shouldn't we add all type origins from the table to the cache considering that we are not going to re-process origin table upon seeing the same `ModuleId`?

Reply:
I don't think we need to do that -- the case that you are thinking of (not re-processing an origin table) only applies to the origin table of the link context, which is different from this path, which handles requests for types that may be defined in a transitive dependency.  If we get multiple requests for such types, we will look again for the type.

It's a bit convoluted, but here's how we ended up here:

 - Initially we had only the implementation you see in this function, which relocates the package using the linkage table, loads the package at the relocated address and scans its type origin table.
 - However, this does not work for newly published packages, because they will not exist in the store when you call `get_package`, so for those we need to pre-populate the type origin table -- so we introduce `type_origin_cache` for that purpose.
 - Because the cache doesn't get stale when we change link context, we decide to keep it there, and because it's there, we might as well populate it when we do the scan, but technically we only need to populate it for the linkage context, when it comes from an unpublished package.
 
My rationale for proactively loading all type origin entries from the linkage table, but only on-demand for transitive dependencies is that it felt like there was a greater chance that we would touch a large number of the linkage context's types than a large number of any given dependencies' types so the risk of overfetch felt smaller.